File: src/world/shipparts/gun.lua
local Shot = require("world/shot")
local Timer = require("timer")

local Gun = class()

function Gun:__create()
	self.charged = true
	self.rechargeTimer = Timer(0.5)

	return self
end

function Gun.process(orders)
	local shoot = false
	for _, order in ipairs(orders) do
		if order == "shoot" then shoot = true end
	end
	return shoot
end

function Gun:update(inputs, location)
	if not self.charged then
		if self.rechargeTimer:ready(inputs.dt) then
			self.charged = true
		end
	else
		if inputs.controls.gun then
			-- Check if there is a part one block infront of the gun.
			local getPart = inputs.getPart
			if not getPart(location, {0, 1}) then
				self.charged = false
				-- Spawn Shot
				return {Shot, {0, 0, 1}}
			end
		end
	end
end

return Gun

File: src/world/shipparts/gun.rs
use crate::timer::Timer;
use crate::world::types::{Location, Module, ModuleInputs, WorldEvent};
use mlua::prelude::{LuaTable, LuaValue};
use mlua::{Lua, Result, ToLua, UserData, UserDataMethods};

fn process(orders: Vec<String>) -> bool {
    orders.iter().any(|order| order == "shoot")
}

struct Gun {
    charged: bool,
    recharge_timer: Timer,
}

impl Module for Gun {
    fn update(&mut self, _: &Lua, inputs: ModuleInputs, location: Location) -> Option<WorldEvent> {
        if !self.charged {
            if self.recharge_timer.ready(inputs.dt) {
                self.charged = true;
            }
            None
        } else if inputs.controls.gun {
            // Check if there is a part one block in front of the gun.
            let part_exists = match inputs.get_part.call::<_, bool>((location, [0, 1])) {
                Ok(part) => part,
                Err(error) => panic!("failed to look up part: {:?}", error),
            };
            if part_exists {
                None
            } else {
                self.charged = false;
                Some(WorldEvent {
                    event_type: "shot".to_string(),
                    location: [0.0, 0.0, 1.0],
                    data: 1.0,
                })
            }
        } else {
            None
        }
    }
}

impl UserData for Gun {
    fn add_methods<'lua, M: UserDataMethods<'lua, Self>>(methods: &mut M) {
        methods.add_method_mut(
            "update",
            |lua, this, (inputs, location): (ModuleInputs, Location)| {
                Ok(Gun::update(this, lua, inputs, location))
            },
        );
    }
}

pub fn lua_module(lua: &Lua) -> Result<LuaTable> {
    let exports = lua.create_table()?;
    exports.set(
        "process",
        lua.create_function(|_, orders: Vec<String>| Ok(process(orders)))?,
    )?;

    let metatable = lua.create_table()?;
    metatable.set(
        "__call",
        lua.create_function(|lua, _: LuaValue| {
            Gun {
                charged: true,
                recharge_timer: Timer {
                    limit: 0.5,
                    time: 0.5,
                },
            }
            .to_lua(lua)
        })?,
    )?;
    exports.set_metatable(Some(metatable));

    Ok(exports)
}
